src/checker.rs

```diff
@@ -690,6 +697,128 @@ pub(crate) enum CheckerInst {
     },
 }
 
+#[derive(Debug)]
+struct DebugLocationEntry {
+    vreg: VReg,
+    alloc: Allocation,
+    label: u32,
+}
+
+#[derive(Debug)]
+struct DebugLocations {
+    expected_vreg_locations: FxHashMap<(VReg, (ProgPoint, ProgPoint), u32), Allocation>,
+}
+
+impl DebugLocations {
+    fn ranges_overlaps(
+        (start_point, end_point): (ProgPoint, ProgPoint),
+        (start_inst, end_inst): (Inst, Inst),
+    ) -> Option<(ProgPoint, ProgPoint)> {
+        if end_inst <= start_point.inst() || start_inst >= end_point.inst() {
+            None
+        } else {
+            let point0 = if start_point.inst() >= start_inst {
+                start_point
+            } else {
+                ProgPoint::before(start_inst)
+            };
+            let point1 = if end_point.inst() < end_inst {
+                end_point
+            } else {
+                ProgPoint::before(end_inst)
+            };
+            Some((point0, point1))
+        }
+    }
+
+    fn new<F: Function>(f: &F, output: &Output) -> Self {
+        let mut expected_vreg_locations = FxHashMap::default();
+        for (label, start_point, end_point, alloc) in &output.debug_locations {
+            for (vreg, start_inst, end_inst, in_label) in f.debug_value_labels() {
+                if in_label != label {
+                    continue;
+                }
+                if let Some(range) =
+                    Self::ranges_overlaps((*start_point, *end_point), (*start_inst, *end_inst))
+                {
+                    expected_vreg_locations.insert((*vreg, range, *label), *alloc);
+                }
+            }
+        }
+        Self {
+            expected_vreg_locations,
+        }
+    }
+
+    fn points_covers_inst(
+        &self,
+        inst: Inst,
+        start_point: ProgPoint,
+        end_point: ProgPoint,
+    ) -> (bool, bool) {
+        let start_inst = start_point.inst();
+        let end_inst = end_point.inst();
+        if inst > start_inst && inst < end_inst {
+            return (true, true);
+        }
+        if inst == start_inst && start_point.pos() == InstPosition::Before {
+            return (true, true);
+        }
+        // Don't check for the case where inst == start and pos == after
+        // because it may be edit instructions after inst that are responsible
+        // for moving the vreg into the expected allocation.
+        if inst == end_inst && end_point.pos() == InstPosition::After {
+            return (true, false);
+        }
+        (false, false)
+    }
+
+    fn entries_covering(&self, inst: Inst) -> Vec<(bool, bool, DebugLocationEntry)> {
+        let mut entries = vec![];
+        for entry in self.expected_vreg_locations.keys().copied() {
```

Comment:
Here is another place I'm a bit worried about checker-time blowup: we're iterating over all vreg locations (which is O(|num vregs| * |program size|)) once per `inst`, so if |vregs| ~= |insts|, this is approximately cubic overall, right?